02_knight_and_princess/roman_kuprii/main.cpp

```diff
@@ -0,0 +1,187 @@
+    res += to_string(connected_vertices[0]);
+    res += ", ";
+    res += to_string(connected_vertices[1]);
+    res += "] }";
+    return res;
+  }
+};
+
+bool is_edge_id_included(const EdgeId& id, const vector<EdgeId>& edge_ids) {
+  for (const auto& edge_id : edge_ids)
+    if (id == edge_id)
+      return true;
+  return false;
+}
+
+struct Vertex {
+ public:
+  const VertexId id = INVALID_ID;
+
+  explicit Vertex(const VertexId& _id) : id(_id) {}
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id) + ", \"edge_ids\": [";
+    for (const auto& edge_id : edges_ids_) {
+      res += to_string(edge_id);
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += "] }";
+    return res;
+  }
+
+  void add_edge_id(const EdgeId& _id) {
+    assert(!is_edge_id_included(_id, edges_ids_));
+    edges_ids_.push_back(_id);
+  }
+
+  vector<EdgeId> get_edges_ids() const { return edges_ids_; }
+
+ private:
+  std::vector<EdgeId> edges_ids_;
+};
+
+bool is_vertex_id_included(const std::vector<Vertex>& vertices,
+                           const VertexId& id) {
+  for (const auto& vert : vertices)
+    if (vert.id == id)
+      return (true);
+  return false;
+}
+
+class Graph {
+ public:
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"vertices\": [ ";
+    for (const auto& v_it : vertices_) {
+      res += v_it.to_json();
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += " ], \"edges\": [ ";
+    for (const auto& e_it : edges_) {
+      res += e_it.to_json();
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += " ] }\n";
+    return res;
+  }
+
+  void add_vertex() { vertices_.emplace_back(get_next_vertex_id()); }
+
+  bool is_connected(const VertexId& from_vertex_id,
+                    const VertexId& to_vertex_id) const {
+    vector<EdgeId> from_vertex_edges_ids =
+        vertices_[from_vertex_id].get_edges_ids();
+    vector<EdgeId> to_vertex_edges_ids =
+        vertices_[to_vertex_id].get_edges_ids();
+    for (const auto& from_vertex_edge_id : from_vertex_edges_ids) {
+      for (const auto& to_vertex_edge_id : to_vertex_edges_ids) {
+        if (from_vertex_edge_id == to_vertex_edge_id)
+          return true;
+      }
+    }
+    return false;
+  }
+
+  void connect_vertices(const VertexId& from_vertex_id,
+                        const VertexId& to_vertex_id) {
+    // check if vertices exist
+    assert(is_vertex_id_included(vertices_, from_vertex_id));
+    assert(is_vertex_id_included(vertices_, to_vertex_id));
+
+    // check if they are not connected
+    assert(!is_connected(from_vertex_id, to_vertex_id));
+
+    const EdgeId id = get_next_edge_id();
+    edges_.emplace_back(from_vertex_id, to_vertex_id, id);
+
+    // add information into Verex structure
+    vertices_[from_vertex_id].add_edge_id(id);
+    vertices_[to_vertex_id].add_edge_id(id);
```

Comment:
nit: Надежней было бы так:
```suggestion
    const auto&  new_edge = edges_.emplace_back(from_vertex_id, to_vertex_id, get_next_edge_id());

    // add information into Vertex structure
    vertices_[from_vertex_id].add_edge_id(new_edge.id);
    vertices_[to_vertex_id].add_edge_id(new_edge.id);
```